File: progresslibrary/src/main/java/com/zhongjh/progresslibrary/entity/MultiMediaView.java
package com.zhongjh.progresslibrary.entity;

import android.view.View;

import com.zhongjh.common.entity.MultiMedia;
import com.zhongjh.common.enums.MultimediaTypes;
import com.zhongjh.progresslibrary.widget.MaskProgressView;
import com.zhongjh.progresslibrary.widget.PlayProgressView;

/**
 * 多媒体实体类,包含着view
 *
 * @author zhongjh
 * @date 2019/1/22
 */
public class MultiMediaView extends MultiMedia {


    private final static int FULL_PERCENT = 100;

    /**
     * 绑定子view,包含其他所有控件（显示view,删除view）
     */
    private View itemView;
    /**
     * 绑定音频View
     */
    private PlayProgressView playProgressView;
    /**
     * 绑定子view: 用于显示图片、视频的view
     */
    private MaskProgressView maskProgressView;
    /**
     * 是否进行上传动作
      */
    private boolean isUploading;

    public MultiMediaView(@MultimediaTypes int multiMediaState) {
        setType(multiMediaState);
    }

    public void setItemView(View itemView) {
        this.itemView = itemView;
    }

    public View getItemView() {
        return itemView;
    }

    public void setMaskProgressView(MaskProgressView maskProgressView) {
        this.maskProgressView = maskProgressView;
    }

    public void setPlayProgressView(PlayProgressView playProgressView) {
        this.playProgressView = playProgressView;
    }

    public MaskProgressView getMaskProgressView() {
        return maskProgressView;
    }

    public PlayProgressView getPlayProgressView() {
        return playProgressView;
    }

    /**
     * 给予进度，根据类型设置相应进度动作
     */
    public void setPercentage(int percent) {
        if (getType() == MultimediaTypes.PICTURE || getType() == MultimediaTypes.VIDEO) {
            this.maskProgressView.setPercentage(percent);
        } else if (getType() == MultimediaTypes.AUDIO) {
            // 隐藏显示音频的设置一系列动作
            this.playProgressView.mViewHolder.numberProgressBar.setProgress(percent);
            if (percent == FULL_PERCENT) {
                this.playProgressView.audioUploadCompleted();
            }
        }
    }

    public boolean isUploading() {
        return isUploading;
    }

    public void setUploading(boolean uploading) {
        isUploading = uploading;
    }

    @Override
    public boolean equals(Object obj) {
        // 父类已重写
        return super.equals(obj);
    }

    @Override
    public int hashCode() {
        // 父类已重写
        return super.hashCode();
    }
}

File: progresslibrary/src/main/java/com/zhongjh/progresslibrary/entity/MultiMediaView.kt
package com.zhongjh.progresslibrary.entity

import android.os.Parcel
import android.os.Parcelable
import android.view.View
import com.zhongjh.common.entity.LocalMedia
import com.zhongjh.progresslibrary.widget.MaskProgressView
import com.zhongjh.progresslibrary.widget.PlayProgressView
import kotlinx.android.parcel.Parcelize

/**
 * 多媒体实体类,包含着view
 *
 * @author zhongjh
 * @date 2021/12/13
 */
class MultiMediaView : LocalMedia, Parcelable {



    /**
     * 绑定子view,包含其他所有控件（显示view,删除view）
     */
    lateinit var itemView: View

    /**
     * 绑定音频View
     */
    lateinit var playProgressView: PlayProgressView

    /**
     * 绑定子view: 用于显示图片、视频的view
     */
    lateinit var maskProgressView: MaskProgressView

    /**
     * 用于区分，因为九宫数据是允许选择重复的
     */
    var multiMediaId: Long = 0

    /**
     * 在线网址
     */
    var url: String? = null

    /**
     * 是否进行上传动作
     */
    var isUploading = false

    constructor() : super()

    constructor(parcel: Parcel) : super(parcel) {
        url = parcel.readString()
        isUploading = parcel.readByte() != 0.toByte()
    }

    constructor(localMedia: LocalMedia) : super(localMedia)

    constructor(mimeType: String) {
        this.mimeType = mimeType
    }

    /**
     * 给予进度，根据类型设置相应进度动作
     */
    fun setPercentage(percent: Int) {
        if (isImageOrGif() || isVideo()) {
            this.maskProgressView.setPercentage(percent)
        } else if (isAudio()) {
            // 隐藏显示音频的设置一系列动作
            playProgressView.mViewHolder.numberProgressBar.progress = percent
            if (percent == FULL_PERCENT) {
                playProgressView.audioUploadCompleted()
            }
        }
    }

    override fun writeToParcel(parcel: Parcel, flags: Int) {
        super.writeToParcel(parcel, flags)
        parcel.writeLong(multiMediaId)
        parcel.writeString(url)
        parcel.writeByte(if (isUploading) 1 else 0)
    }

    override fun describeContents(): Int {
        return 0
    }

    companion object CREATOR : Parcelable.Creator<MultiMediaView> {

        private const val FULL_PERCENT = 100

        override fun createFromParcel(parcel: Parcel): MultiMediaView {
            return MultiMediaView(parcel)
        }

        override fun newArray(size: Int): Array<MultiMediaView?> {
            return arrayOfNulls(size)
        }
    }

}